dali/python/nvidia/dali/plugin/pytorch/__init__.py

```diff
@@ -795,3 +853,266 @@ def __next__(self) -> List[Dict[str, torch.Tensor]]:
+    """
+    assert len(pipe_out) > 0
+    first_elem = pipe_out[0]
+    inputs = [[] for idx in range(len(first_elem.inputs))]
+    for elem in pipe_out:
+        assert first_elem.dali_proxy == elem.dali_proxy
+        for idx, input_ref in enumerate(elem.inputs):
+            inputs[idx].append(input_ref)
+    return first_elem.dali_proxy.schedule_batch(inputs)
+
+
+def custom_collate(batch):
+    """
+    Subscribe a special collate function for PipelineRunRef, that handles the scheduling of the iteration
+    on the fly
+    """
+    collate_fn_map = data._utils.collate.default_collate_fn_map
+    collate_fn_map.update({DALIProxy.PipelineRunRef: collate_pipeline_run_ref_fn})
+    return collate(batch, collate_fn_map=collate_fn_map)
+
+
+def flatten_tuple(nested_tuple):
+    """
+    Flattens a nested tuple
+    """
+    flat_list = []
+
+    def _flatten(t):
+        for item in t:
+            if isinstance(item, tuple):
+                _flatten(item)
+            else:
+                flat_list.append(item)
+
+    _flatten(nested_tuple)
+    return tuple(flat_list)
+
+
+class DALIMultiProcessingDataLoaderIter(_MultiProcessingDataLoaderIter):
+    """
+    Data loader iterator used by the DALI proxy data loader
+    """
+    def __init__(self, loader):
+        super().__init__(loader)
+        self.loader = loader
+
+    def _next_data(self):
+        data = super()._next_data()
+        if not hasattr(data, "__iter__"):
+            print(
+                "Warning: Non iterable returned from dataloader. Please "
+                " review the code, since it usually indicates a bug in a pipeline."
+            )
+            data = [data]
+        for data_idx, data_elem in enumerate(data):
+            # If loader returns a dictionary the iterator iterates over its keys.
+            # We need to access a value. Probably need to address more casess.
+            if isinstance(data, dict):
+                if isinstance(data[data_elem], DALIProxy.PipelineOutputRef):
+                    data[data_elem] = self.loader.get_outputs(data[data_elem].info)
+            if isinstance(data_elem, DALIProxy.PipelineOutputRef):
+                data[data_idx] = self.loader.get_outputs(data_elem.info)
+        return data
+
+
+class DALIDataLoader(DataLoader):
+    """
+    DALI data loader to be used in the main loop, which runs the DALI pipeline doing the processing
+    asynchronously with regards to the training.
+    """
+    def __init__(self, pipe, dali_proxy, *args, **kwargs):
+        if "collate_fn" in kwargs and kwargs["collate_fn"] is not None:
+            print(
+                "Warning: Make sure to handle DALIProxy.PipelineRunRef when providing"
+                " a custom collate_fn"
+            )
+        else:
+            kwargs["collate_fn"] = custom_collate
+        super().__init__(*args, **kwargs)
+        self.pipe = pipe
+        self.dali_proxy = dali_proxy
+        self.t = None
```

Comment:
Nitpick: At first I thought that `t` is some kind of timestamp. How about changing it to `thread`?